### PR TITLE
feat: display chat events

### PR DESCRIPTION
**Pull request by thucpn**
## Summary by CodeRabbit

- **New Features**
  - Introduced enhanced content generation for chat functionalities, handling both text and events.
  - Added new callback event handling classes for improved streaming data management.
  - Implemented collapsible chat events components in the UI, enhancing user interaction.
  
- **Enhancements**
  - Updated streaming data functions and response handling for better performance and flexibility.
  
- **Dependencies**
  - Added `aiostream` and `@radix-ui/react-collapsible` to manage asynchronous streams and UI collapsibility, respectively.

### CODE REVIEW

File: templates/types/streaming/nextjs/app/api/chat/route.ts [hunk 1]
```diff
@@ -54,6 +60,17 @@ export async function POST(request: NextRequest) {
       data?.imageUrl,
     );
 
+    // Init Vercel AI StreamData
+    const vercelStreamData = new StreamData();
+
+    // Setup callback for streaming data before chatting
+    Settings.callbackManager = new CallbackManager({
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/api/chat/route.ts:67]: @thucpn can you try this (`onRetrieve` will be deprecated soon)
```suggestion
    Settings.callbackManager.on("retrieve"
```

File: templates/types/streaming/nextjs/app/api/chat/route.ts [hunk 2]
```diff
@@ -62,14 +79,14 @@ export async function POST(request: NextRequest) {
     });
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/api/chat/route.ts:80]: @thucpn currently, TS doesn't have an event for starting the retrieval process (I will add it later), but can you send first a 

`appendEventData(vercelStreamData, 'Retrieving context for query ${query_str}');`

and change the text above to

`appendEventData(vercelStreamData, 'Retrieved ${nodes} sources to use as context for the query');`

Also please make sure that your UI supports displaying multiple events

File: templates/types/streaming/nextjs/app/api/chat/route.ts [hunk 3]
```diff
@@ -62,13 +57,18 @@ export async function POST(request: NextRequest) {
...
-    Settings.callbackManager = new CallbackManager({
-      onRetrieve: ({ query, nodes }) => {
-        const eventTitle = `Retrieved ${nodes.length} nodes for query: '${query}'`;
-        appendEventData(vercelStreamData, eventTitle);
-      },
+    Settings.callbackManager.on("retrieve", (data) => {
+      const { query, nodes } = data.detail;
+      appendEventData(
+        vercelStreamData,
+        `Retrieved ${nodes.length} sources to use as context for the query: '${query}'`,
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/api/chat/route.ts:70]: same as for express

File: templates/types/streaming/fastapi/app/api/routers/chat.py [hunk 4]
```diff
@@ -116,16 +116,10 @@ async def _event_generator():
...
-        async for item in asyncio.as_completed([_text_generator(), _event_generator()]):
-            async for value in item:
-                # If client closes connection, stop sending events
-                if await request.is_disconnected():
-                    break
-                yield value
+        combine = stream.merge(_text_generator(), _event_generator())
+        async with combine.stream() as streamer:
+            async for item in streamer:
+                yield item
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/api/routers/chat.py:122]: @leehuwuj i guess this is missing:
```
# If client closes connection, stop sending events
                if await request.is_disconnected():
                    break
```

Comment (leehuwuj) [reply templates/types/streaming/fastapi/app/api/routers/chat.py:124]: Great catch! I just updated.

File: templates/types/streaming/fastapi/app/api/routers/chat.py [hunk 5]
```diff
@@ -92,15 +94,34 @@ async def chat(
...
         # Yield the text response
-        async for token in response.async_response_gen():
-            # If client closes connection, stop sending events
-            if await request.is_disconnected():
-                break
-            yield VercelStreamResponse.convert_text(token)
+        async def _text_generator():
+            async for token in response.async_response_gen():
+                yield VercelStreamResponse.convert_text(token)
+            # TODO: ideally we don't need is_done and we just consume till _text_generator is finished below
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/api/routers/chat.py:106]: @leehuwuj can you have a look at this

**Pull request merged**